### PR TITLE
DEPR: require SparseDtype.fill_value be compatible with SparseDtype.subtype

### LINKED ISSUE #23124: Require the dtype of SparseArray.fill_value and sp_values.dtype to match
This has confused me for a while, but we apparently (sometimes?) allow a `fill_value` whose dtype is not the same as `sp_values.dtype`

```python
In [20]: a = pd.SparseArray([1, 2, 3], fill_value=1.0)

In [21]: a
Out[21]:
[1.0, 2, 3]
Fill: 1.0
IntIndex
Indices: array([1, 2], dtype=int32)

In [22]: a.sp_values.dtype
Out[22]: dtype('int64')
```

This can lead to confusing behavior when doing operations.

I suspect a primary motivation was supporting sparse integer values with `NaN` for a fill value. We should investigate what's tested, part of the API, and useful to users.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #23124 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/arrays/sparse/test_dtype.py [hunk 1]
```diff
@@ -67,15 +68,21 @@ def test_nans_equal():
     assert b == a
 
 
-@pytest.mark.parametrize(
-    "a, b",
-    [
+with warnings.catch_warnings():
+    warnings.filterwarnings("ignore")
```
Comment (mroeschke) [inline pandas/tests/arrays/sparse/test_dtype.py:72]: Could you make this filter more specific?

File: pandas/core/arrays/sparse/dtype.py [hunk 2]
```diff
@@ -161,18 +166,41 @@ def _check_fill_value(self):
...
+            if not is_valid_na_for_dtype(val, self.subtype):
+                warnings.warn(
+                    "Allowing arbitrary scalar fill_value in SparseDtype is "
+                    "deprecated. In a future version, the fill_value must be "
+                    "a valid value for the SparseDtype.subtype.",
+                    FutureWarning,
+                    stacklevel=find_stack_level(),
+                )
+        elif isinstance(self.subtype, CategoricalDtype):
+            # TODO: is this even supported?  It is reached in
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/dtype.py:182]: in principle i guess, but i think this case existing at all is wrong. SparseArray._sparse_values has to be np.ndarray, and the dtype is initialized with `self._dtype = SparseDtype(sparse_values.dtype, fill_value)`

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/dtype.py:182]: will make a separate PR to disallow this case

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/dtype.py:182]: #53160 disallows this case

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/dtype.py:185]: should no longer be an issue now that #53160 is merged

**Review by mroeschke — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

File: pandas/core/arrays/sparse/dtype.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/dtype.py:185]: woops, should have updated to remove this check.  will do in a follow-up

**Pull request merged**